chunky/src/java/se/llbit/math/NodeBasedOctree.java

```diff
@@ -284,138 +289,76 @@ public boolean enterBlock(Scene scene, Ray ray, BlockPalette palette) {
           tNear = t;
           nz = -1;
           nx = ny = 0;
         }
       }
 
-      ray.o.scaleAdd(tNear, d);
+      ray.o.scaleAdd(tNear, ray.d);
       ray.n.set(nx, ny, nz);
       ray.distance += tNear;
-      tNear = Double.POSITIVE_INFINITY;
     }
   }
 
   @Override
   public boolean exitWater(Scene scene, Ray ray, BlockPalette palette) {
-    int level;
-    Octree.Node node;
-    boolean first = true;
-
-    int lx, ly, lz;
-    int x, y, z;
-    int nx = 0, ny = 0, nz = 0;
-    double tNear = Double.POSITIVE_INFINITY;
-    double t;
-    Vector3 d = ray.d;
+    if (!isInside(ray.o) && !enterOctree(ray))
+      return false;
 
     while (true) {
+      int x = (int) QuickMath.floor(ray.o.x + ray.d.x * Ray.OFFSET);
+      int y = (int) QuickMath.floor(ray.o.y + ray.d.y * Ray.OFFSET);
+      int z = (int) QuickMath.floor(ray.o.z + ray.d.z * Ray.OFFSET);
 
-      x = (int) QuickMath.floor(ray.o.x + d.x * Ray.OFFSET);
-      y = (int) QuickMath.floor(ray.o.y + d.y * Ray.OFFSET);
-      z = (int) QuickMath.floor(ray.o.z + d.z * Ray.OFFSET);
-
-      node = root;
-      level = depth;
-      lx = x >>> level;
-      ly = y >>> level;
-      lz = z >>> level;
-
-      if (lx != 0 || ly != 0 || lz != 0) {
-
-        // only check octree intersection if this is the first iteration
-        if (first) {
-          // test if it is entering the octree
-          t = -ray.o.x / d.x;
-          if (t > Ray.EPSILON) {
-            tNear = t;
-            nx = 1;
-            ny = nz = 0;
-          }
-          t = ((1 << level) - ray.o.x) / d.x;
-          if (t < tNear && t > Ray.EPSILON) {
-            tNear = t;
-            nx = -1;
-            ny = nz = 0;
-          }
-          t = -ray.o.y / d.y;
-          if (t < tNear && t > Ray.EPSILON) {
-            tNear = t;
-            ny = 1;
-            nx = nz = 0;
-          }
-          t = ((1 << level) - ray.o.y) / d.y;
-          if (t < tNear && t > Ray.EPSILON) {
-            tNear = t;
-            ny = -1;
-            nx = nz = 0;
-          }
-          t = -ray.o.z / d.z;
-          if (t < tNear && t > Ray.EPSILON) {
-            tNear = t;
-            nz = 1;
-            nx = ny = 0;
-          }
-          t = ((1 << level) - ray.o.z) / d.z;
-          if (t < tNear && t > Ray.EPSILON) {
-            tNear = t;
-            nz = -1;
-            nx = ny = 0;
-          }
+      int lx = x >>> depth;
+      int ly = y >>> depth;
+      int lz = z >>> depth;
 
-          if (tNear < Double.MAX_VALUE) {
-            ray.o.scaleAdd(tNear, d);
-            ray.n.set(nx, ny, nz);
-            ray.distance += tNear;
-            tNear = Double.POSITIVE_INFINITY;
-            continue;
-          } else {
-            return false;// outside of octree!
-          }
-        } else {
-          return false;// outside of octree!
-        }
-      }
-
-      first = false;
+      if (lx != 0 || ly != 0 || lz != 0)
+        return false;
```

Comment:
A comment that notes that the ray is outside of the octree here would be good. Explaining how the octree intersection works by a few comments would improve this code.

Reply:
I added a few comments in the intersection functions to try and give an idea of how they work, but I don't fully understand them (especially the part that decides when to terminate the loop).